cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -136,14 +153,50 @@ protected void addPartition(PartitionKey key, String path, boolean addToExplore)
       put.add(Bytes.add(FIELD_PREFIX, Bytes.toBytes(entry.getKey())), // "f.<field name>"
               Bytes.toBytes(entry.getValue().toString()));            // "<string rep. of value>"
     }
+
+    addMetadataToPut(metadata, put);
+
     partitionsTable.put(put);
     partitionsAddedInSameTx.put(path, key);
-    if (addToExplore) {
+
+    if (explorable) {
       addPartitionToExplore(key, path);
       // TODO: make DDL operations transactional [CDAP-1393]
     }
   }
 
+  @Override
+  public void addMetadata(PartitionKey key, String metadataKey, String metadataValue) {
+    addMetadata(key, ImmutableMap.of(metadataKey, metadataValue));
+  }
+
+  @Override
+  public void addMetadata(PartitionKey key, Map<String, String> metadata) {
+    final byte[] rowKey = generateRowKey(key, partitioning);
+    Row row = partitionsTable.get(rowKey);
+    if (row == null) {
+      throw new DataSetException(String.format("Dataset '%s' does not have a partition for key: %s", getName(), key));
+    }
+
+    // ensure that none of the entries already exist in the metadata
+    for (Map.Entry<String, String> metadataEntry : metadata.entrySet()) {
+      String metadataKey = metadataEntry.getKey();
+      byte[] columnKey = columnKeyFromMetadataKey(metadataKey);
+      Preconditions.checkArgument(row.get(columnKey) == null, "Entry already exists for metadata key: {}", metadataKey);
```

Comment:
Actually, when you add a partition with a partition key that already exists, it throws a DatasetException. For consistency (since this is a similar situation), it should throw the same exception, right?